services-api/src/main/java/io/scalecube/services/ServiceInfo.java

```diff
@@ -85,26 +85,82 @@ private Builder(Object serviceInstance) {
       this.serviceInstance = serviceInstance;
     }
 
+    /**
+     * Add tag for service info.
+     *
+     * @param key tag name.
+     * @param value tag value.
+     * @return current builder's state.
+     */
     public Builder tag(String key, String value) {
       tags.put(key, value);
       return this;
     }
 
+    /**
+     * Set up {@link ServiceProviderErrorMapper}.
+     *
+     * @param errorMapper error mapper.
+     * @return current builder's state.
+     */
     public Builder errorMapper(ServiceProviderErrorMapper errorMapper) {
       this.errorMapper = errorMapper;
       return this;
     }
 
+    /**
+     * Set up {@link ServiceMessageDataDecoder}.
+     *
+     * @param dataDecoder data decoder.
+     * @return current builder's state.
+     */
     public Builder dataDecoder(ServiceMessageDataDecoder dataDecoder) {
       this.dataDecoder = dataDecoder;
       return this;
     }
 
+    /**
+     * Set up {@link ServiceProviderErrorMapper} if it hasn't been set up before.
+     *
+     * @param errorMapper error mapper.
+     * @return current builder's state.
+     */
+    public Builder errorMapperIfAbsent(ServiceProviderErrorMapper errorMapper) {
```

Comment:
Didn't get idea of those `.***IfAbsent`  methods. Why u need them? Irregardless of `servicesProvider` type u go through `.registerInMethodRegistry()` anyway which, by turn, sets corresponding dependencies.

Reply:
because this looks pretty, see:

```java
.errorMapper(Optional.ofNullable(serviceInfo.authenticator()).orElse(this.errorMapper))
```
vs

```java
.errorMapperIfAbsent(this.errorMapper)
```

similar to Map#putIfAbsent.